docs/source/upgrade.rst

```diff
@@ -0,0 +1,759 @@
+
+.. admonition:: Example
+
+   For example, the following chain initialization code
+
+   .. code-block:: py
+
+      # Chainer v1
+      class MyMLP(chainer.Chain):
+          def __init__(self):
+              super(MyMLP, self).__init__(
+                  layer1=L.Linear(None, 20),
+                  layer2=L.Linear(None, 30),
+              )
+          ...
+
+   is recommended to be updated as follows.
+
+   .. code-block:: py
+
+      # Chainer v2
+      class MyMLP(chainer.Chain):
+          def __init__(self):
+              super(MyMLP, self).__init__()
+              with self.init_scope():
+                  self.layer1 = L.Linear(20)
+                  self.layer2 = L.Linear(30)
+
+   Note that this example also demonstrates the use of new APIs with :ref:`the omitted input size <update-omit-input-size>`, explained below.
+
+.. note::
+
+   To keep a :class:`Link` object as an attribute without registration, you can set the attribute without using the ``with self.init_scope():`` block.
+
+.. _update-omit-input-size:
+
+The input-size placeholder of links are made optional
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, the input size of many links, including :class:`~links.Linear` and :class:`~links.Convolution2D`, is made optional.
+In Chainer v1, we had to use ``None`` as the placeholder to specify that the input size should be determined at the first iteration.
+The placeholder can also be used in Chainer v2, although it is easier to just omit the input size.
+
+See the previous item for the example of omitting the input size of :class:`~links.Linear`.
+The following links currently support the omitted input size.
+
+- :class:`~links.Convolution2D`
+- :class:`~links.Deconvolution2D`
+- :class:`~links.DilatedConvolution2D`
+- :class:`~links.Linear`
+- :class:`~links.LSTM`
+- :class:`~links.MLPConvolution2D`
+- :class:`~links.StatelessLSTM`
+
+
+Optimizer
+---------
+
+Deprecated methods of Optimizer are removed
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following methods are removed from :class:`Optimizer`.
+These methods have been already deprecated in the past versions.
+**If you are using these methods, you have to update your code.**
+
+- ``zero_grads``: use :meth:`Link.zerograds` instead.
+- ``compute_grads_norm``: you can compute the gradient norm by iterating the list of parameters by :meth:`Link.params`.
+- ``clip_grads``: use :class:`~optimizer.GradientClipping` instead.
+- ``weight_decay``: use :class:`~optimizer.WeightDecay` instead.
+- ``accumulate_grads``: use :meth:`Link.addgrads`` instead.
+
+.. _upgrade-gradient-method-cleargrads:
+
+GradientMethod uses Link.cleargrads instead of Link.zerograds by default
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, :class:`GradientMethod` clears the gradient before running backprop by :meth:`Link.cleargrads`.
```

Comment:
Perhaps most users are not aware that most of commonly-used optimizers are a subclass of `GradientMethod` unless they look at the code or carefully read the document. So, I suggest that this change is affects most of optimizers like `SGD`, `Adam` and so on.